### PR TITLE
Fix memory leak in wat2wasm function

### PR DESCRIPTION
**Pull request by MarkMcCaskey**
Alternative to #1853 ; 

Found with @nlewycky 

# Review

- [x] Add a short description of the the change to the CHANGELOG.md file

### CODE REVIEW

File: lib/c-api/src/wasm_c_api/wat.rs [hunk 1]
```diff
@@ -6,11 +6,20 @@ use super::types::wasm_byte_vec_t;
...
+    let wat: &[u8] = match wat.into_slice() {
+        Some(v) => v,
+        _ => return,
+    };
+    let result: wasm_byte_vec_t = match wasmer::wat2wasm(wat) {
+        Ok(val) => val.into_owned().into(),
+        Err(err) => {
+            crate::error::update_last_error(err);
+            return;
+        }
```
Comment (nlewycky) [inline lib/c-api/src/wasm_c_api/wat.rs:19]: How does the user know whether `wat2wasm` succeeded? Is there a way to initialize a `wasm_byte_vec_t` to "clear" which the user can then test for? Or does it always check for errors with a separate "did the last thing fail" call after every call to `wat2wasm`?

Comment (MarkMcCaskey) [reply lib/c-api/src/wasm_c_api/wat.rs:19]: That's a fair point, in this case I'd assume the user would initialize to null prior and check for that, but perhaps it makes more sense for us to zero the fields for them on failure

Comment (nlewycky) [reply lib/c-api/src/wasm_c_api/wat.rs:19]: You can also use the return value if you like.

File: lib/c-api/tests/wasm_c_api/test-wat2wasm.c [hunk 2]
```diff
@@ -20,31 +20,32 @@ int main(int argc, const char* argv[]) {
     .data = "(module)",
     .size = 8,
   };
-  wasm_byte_vec_t *wasm = wat2wasm(&wat);
+  wasm_byte_vec_t wasm;
+  wat2wasm(&wat, &wasm);
 
-  if (!wasm) {
+  if (!wasm.data) {
     printf("> Error compiler WAT to Wasm!\n");
```
Comment (nlewycky) [inline lib/c-api/tests/wasm_c_api/test-wat2wasm.c:27]:
```suggestion
    printf("> Error compiling WAT to Wasm!\n");
```

File: lib/c-api/wasmer_wasm.h [hunk 3]
```diff
@@ -239,6 +239,6 @@ int wasmer_last_error_message(char *buffer, int length);
  *
  * In case of failure, `wat2wasm` returns `NULL`.
```
Comment (nlewycky) [inline lib/c-api/wasmer_wasm.h:240]:
```suggestion
 * In case of failure, `wat2wasm` fills the `out->data = NULL` and `out->size = 0`.
```

**Review by nlewycky — APPROVED** [review]

File: lib/c-api/wasmer_wasm.h [hunk 4]
```diff
@@ -237,8 +237,8 @@ int wasmer_last_error_message(char *buffer, int length);
  * Parses in-memory bytes as either the WAT format, or a binary Wasm
  * module. This is wasmer-specific.
  *
- * In case of failure, `wat2wasm` returns `NULL`.
+ * In case of failure, `wat2wasm` fills the `out->data = NULL` and `out->size = 0`.
```
Comment (nlewycky) [inline lib/c-api/wasmer_wasm.h]: Grammar here is a little off, but meh.

Comment (MarkMcCaskey) [reply lib/c-api/wasmer_wasm.h]: I fixed it!

**Comment by Hywan** [comment]
:+1:

**Pull request merged**